### PR TITLE
LWG Poll 18: P2091R0 Issues with range access CPOs

### LINKED ISSUE #3720: [2020-02 LWG Poll 18] P2091R0 Issues with range access CPOs
resolves GB275, LWG issues 3258, 3299, 3368

### PR DESCRIPTION
**Pull request by burblebee**
Fixes #3720.
Fixes cplusplus/nbballot#271.

### CODE REVIEW

File: source/ranges.tex [hunk 1]
```diff
@@ -602,54 +680,72 @@
...
+  \tcode{\placeholdernc{decay-copy}(size(E))}
+  with overload resolution performed in the above context.
 
 \item
-  Otherwise, \tcode{\placeholdernc{to-unsigned-like}(ranges::end(E) - ranges::begin(E))}\iref{range.subrange}
-  if it is a valid expression and
-  the types \tcode{I} and \tcode{S} of \tcode{ranges::begin(E)} and
-  \tcode{ranges::end(E)} (respectively) model both
+  Otherwise, if
+  \tcode{\placeholdernc{make-unsigned-like}(ranges::end(t) - ranges::begin(t))}\iref{ranges.syn}
```
Comment (johelegp) [inline source/ranges.tex]:
```suggestion
  \tcode{\placeholdernc{to-unsigned-like}(ranges::end(t) - ranges::begin(t))}\iref{ranges.syn}
```
See https://github.com/cplusplus/draft/issues/3720#issuecomment-587581104.

File: source/ranges.tex [hunk 2]
```diff
@@ -602,54 +680,72 @@
...
+  is a valid expression and
+  the types \tcode{I} and \tcode{S} of \tcode{ranges::begin(t)} and
+  \tcode{ranges::end(t)} (respectively) model both
   \tcode{\libconcept{sized_sentinel_for}<S, I>}\iref{iterator.concept.sizedsentinel} and
-  \tcode{\libconcept{forward_iterator}<I>}.
-  However, \tcode{E} is evaluated only once.
+  \tcode{\libconcept{forward_iterator}<I>},
+  then \tcode{ranges::size(E)} is expres\-sion-equivalent to
+  \tcode{\placeholdernc{make-unsigned-like}(ranges::end(t) - ranges::begin(t))}.
```
Comment (johelegp) [inline source/ranges.tex]:
```suggestion
  \tcode{\placeholdernc{to-unsigned-like}(ranges::end(t) - ranges::begin(t))}.
```
See https://github.com/cplusplus/draft/issues/3720#issuecomment-587581104.

File: source/ranges.tex [hunk 3]
```diff
@@ -692,28 +806,52 @@
...
+  is a valid expression of pointer to object type,
+  \tcode{ranges::data(E)} is expression-equivalent to
+  \tcode{\placeholdernc{decay-copy}(t.data())}.
 
 \item
-  Otherwise, if \tcode{ranges::begin(E)} is a valid expression whose type models
+  Otherwise, if \tcode{ranges::begin(t)}
+  is a valid expression whose type models
   \libconcept{contiguous_iterator},
+  \tcode{ranges::data(E)} is expresssion-equivalent to
```
Comment (jwakely) [inline source/ranges.tex]: The paper contains a typo which is preserved here: "expresssion"

File: source/ranges.tex [hunk 4]
```diff
@@ -511,49 +572,66 @@
...
+  \tcode{\libconcept{sentinel_for}<decltype(ranges::rbegin(E))>}
+  with overload resolution performed in a context in which
+  unqualified lookup for \tcode{rend} finds only the declarations
 \begin{codeblock}
-template<class T> void rend(T&&) = delete;
+void rend(auto&) = delete;
+void rend(const auto&) = delete;
 \end{codeblock}
-  and does not include a declaration of \tcode{ranges::rend}.
+  then \tcode{ranges::rbegin(E)} is expression-equivalent to
```
Comment (CaseyCarter) [inline source/ranges.tex]: Woops - this is copy pasta in the ingress paper; we aren't specifying `ranges::rbegin(E)` here, we're specifying `ranges::rend(E)`.
```suggestion
  then \tcode{ranges::rend(E)} is expression-equivalent to
```

File: source/ranges.tex [hunk 5]
```diff
@@ -602,54 +680,72 @@
...
-    is integer-like
-    with overload resolution performed in a context that includes
-    the declaration:
+\item
+  Otherwise, if \tcode{disable_sized_range<remove_cv_t<T>>}\iref{range.sized}
+  is \tcode{false} and
+  \tcode{\placeholdernc{decay-copy}(t.size())}
+  is a valid expression of integer-like type\iref{iterator.concept.winc},
+  \tcode{ranges::size(E)} is expression-equivalent to
+  \tcode{\placeholdernc{decay-copy}(E.size())}.
```
Comment (CaseyCarter) [inline source/ranges.tex]: Another typo in the ingress paper. What is wrong with these authors?
```suggestion
  \tcode{\placeholdernc{decay-copy}(t.size())}.
```

File: source/ranges.tex [hunk 6]
```diff
@@ -602,54 +680,72 @@
...
+  unqualified lookup for \tcode{size} finds only the declarations
 \begin{codeblock}
-template<class T> void size(T&&) = delete;
+void size(auto&) = delete;
+void size(const auto&) = delete;
 \end{codeblock}
-    and does not include a declaration of \tcode{ranges::size}.
-  \end{itemize}
+  then \tcode{ranges::size(E)} is expression-equivalent to
+  \tcode{\placeholdernc{decay-copy}(size(E))}
```
Comment (CaseyCarter) [inline source/ranges.tex]: Ditto typo.
```suggestion
  \tcode{\placeholdernc{decay-copy}(size(t))}
```

File: source/ranges.tex (hunk 1 above)
Comment (burblebee) [reply source/ranges.tex]: Ok, added as a separate edit

File: source/ranges.tex (hunk 6 above)
Comment (jwakely) [reply source/ranges.tex]: I think we still have the same typo here in `ranges::data`:

https://github.com/cplusplus/draft/blob/caf7428fd6c807f30538042cf9bf1b27b8772b01/source/ranges.tex#L918-L922
We check that `ranges::begin(t)` is valid, and then do `ranges::begin(E)` instead. It should have the same effects, but it's inconsistent and confusing to switch between `t` and `E`.

I've proposed a fix in #4628

**Pull request merged**